### ISSUE TITLE
How to retain any existing refs to children of tooltips?

### ISSUE THREAD

**Issue by LarsEjaas** [floating-ui/floating-ui#1773]
First of all - thanks for a great library and some really great code examples on CodeSandbox 🥳

I am currently using a tooltip build with Floating UI in React but often use `refs` for other functionality for components (primary buttons and anchor tags).
The issue is that I have a hard time finding a solution where I can retain any existing refs to these components when wrapping them with a tooltip.

I have forked the animated tooltip example on `CodeSandbox` and added a bit of code in the `App.tsx` file:
https://codesandbox.io/s/animated-toolbox-zwukoz?file=/src/App.tsx to illustrate the issue.

**In short: Refs added to any children will always return null.**

In the `Tooltip.tsx` file in line 44 we explicitly override any refs on the cloned children with refs from the `useFloating` hook. 

But I have a hard time figuring out the internals of the `react-dom-interactions` package and how you would retain any existing refs and still get to pass a ref to the cloned children to the tooltip? 

I guess you need to use the `useImperativeHandle` hook from React - but not really sure where to add this code?

**Comment by LarsEjaas** [comment]
> The examples fail to show “ref preserving”. You need to merge `children.props.ref` with the reference ref

Thanks a lot for the fast reply @atomic 🙂

Normally I would use a utility function like this to merge refs: 

```
export function mergeRefs<T>(
  refs: Array<React.MutableRefObject<T> | React.LegacyRef<T>>
): React.RefCallback<T> {
  return (value) => {
    refs.forEach((ref) => {
      if (typeof ref === "function") {
        ref(value);
      } else if (ref != null) {
        (ref as React.MutableRefObject<T | null>).current = value;
      }
    });
  };
}
```

However, not sure how to go about it combined with the `getReferenceProps` ?

I would expect something like this to work - (but it doesn't):

```
//...
const mergedRef = mergeRefs([reference, children.props.ref]);

  return (
    <>
      {cloneElement(
        children,
        getReferenceProps({
          ref: mergedRef,
          ...children.props,
        })
      )}
 //...
 ```

**Comment by LarsEjaas** [comment]
> @LarsEjaas try
> 
> ```
> //...
>     const mergeRef = useMemo(
>       () => mergeRefs([reference, children?.ref]),
>       [reference, children?.ref],
>     )
> 
>   return (
>     <>
>       {cloneElement(
>         children,
>         getReferenceProps({
>           ref: mergedRef,
>           ...children.props,
>         })
>       )}
>  //...
> ```

Thanks for the comment @oddvernes. Probably a good call to memorize the `mergeRef` method here. I tried your implementation, but can't make it work. I have to use `children.props.ref` and not `children.ref` though.

But this way I get the ref to the child node as intended, but my tooltip doesn't render. So I can choose between using: 
```
//...
{cloneElement(
        children,
        getReferenceProps({
          ref: reference,
          ...children.props,
//...
```
Where the tooltip works as intended but I lose the ref to the child OR use `ref: children.props.ref` and then have the reference to the child but no tooltip 🤨🤔

**Comment by LarsEjaas** [comment]
> @oddvernes is right, `children.ref` not `children.props.ref`
> 
> I updated the example: https://codesandbox.io/s/winter-tree-wmmffl?file=/src/Tooltip.tsx
> 
> You can see it's preserved here:
> 
> ```js
> export default function App() {
>   const test = useRef<HTMLButtonElement>(null);
> 
>   useEffect(() => {
>     console.log(test.current)
>   }, [])
> 
>   return (
>     <div className="App">
>       <h1>Floating UI Tooltip</h1>
>       <h2>Tooltip</h2>
>       <Tooltip label="My tooltip">
>         <button ref={test}>Hover me</button>
>       </Tooltip>
>       <h2>`framer-motion` animated tooltip</h2>
>       <p>Including pointer "rest" delay</p>
>       <AnimatedTooltip label="My tooltip">
>         <button>Hover me</button>
>       </AnimatedTooltip>
>     </div>
>   );
> }
> ```

Thanks a lot for taking the time to look into this @atomiks 🙏 This is such a big help 🥳
